Mathlib/Data/Nat/Factorization/Multiplicity.lean

```diff
@@ -0,0 +1,349 @@
+/-
+Copyright (c) 2025 Qi Wen Wei. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Qi Wen Wei
+-/
+import Mathlib.Data.Nat.PrimeFin
+import Mathlib.Data.Nat.Factorization.Defs
+import Mathlib.Data.Nat.Log
+import Mathlib.Data.Nat.Digits
+import Mathlib.Algebra.GeomSum
+import Mathlib.Data.Nat.Prime.Defs
+import Mathlib.RingTheory.Multiplicity
+
+
+/-!
+# Natural number factorization
+
+This file contains lemmas about the factorization function (the maximum prime power dividing a
+number) when applied to naturals, in particular calculating it for factorials and binomial
+coefficients.
+
+This is a port of the lemmas from `Data/Nat/Multiplicity` to here,
+re-written in terms of `factorization`
+
+
+## Factorization calculations
+
+* `Nat.Prime.factorization_factorial`: Legendre's Theorem. The factorization of `p` in `n!` is
+  `n / p + ... + n / p ^ b` for any `b` such that `n / p ^ (b + 1) = 0`. See `padicValNat_factorial`
+  for this result stated in the language of `p`-adic valuations and
+  `sub_one_mul_padicValNat_factorial` for a factorization result.
+* `Nat.Prime.factorization_factorial_mul`: The factorization of `p` in `(p * n)!` is `n` more than
+  that of `n!`.
+* `Nat.Prime.factorization_choose`: Kummer's Theorem. The factorization of `p` in `n.choose k`
+   is the number of carries when `k` and `n - k` are added in base `p`.
+   See `padicValNat_choose` for the same result but stated in the language of `p`-adic
+   valuations and `sub_one_mul_padicValNat_choose_eq_sub_sum_digits` for a related result.
+
+## Other declarations
+
+* `Nat.factorization_eq_card_pow_dvd`: The factorization of `m` in `n` is the number of positive
+  natural numbers `i` such that `m ^ i` divides `n`.
+
+## Tags
+
+Legendre, p-adic
+-/
+
+open Finset List Finsupp
+
+namespace Nat
+
+/-- The Factorization of prime `p` in `p` is `1` -/
+@[simp]
+theorem Prime.factorization_self {p : ℕ} (hp : Prime p) : p.factorization p = 1 := by simp [hp]
+
+/-- The factorization of `m` in `n` is the number of positive natural numbers `i` such that `m ^ i`
+divides `n`. Note `m` is prime. This set is expressed by filtering `Ico 1 b` where `b` is any bound
+greater than `log m n`. -/
+theorem factorization_eq_card_pow_dvd₀ {m n b : ℕ} (hm : m ≠ 1)
+  (hm2: m.Prime) (hn : 0 < n) (hb : log m n < b) :
+    n.factorization m = #{i ∈ Ico 1 b | m ^ i ∣ n} :=
+  have fin := Nat.finiteMultiplicity_iff.2 ⟨hm, hn⟩
+  calc
+    n.factorization m = #(Ico 1 <| multiplicity m n + 1) := by {
+      simp only [card_Ico, add_tsub_cancel_right]
+      rw [← multiplicity_eq_factorization]
+      · exact hm2
+      exact not_eq_zero_of_lt hn
+    }
```

Comment:
One stylistic point: typically, we would write
```
have : foo := by 
  tac1
  tac2
```
rather than 
```
have : foo := by  {
  tac1
  tac2
}
```
and so on.